### ISSUE TITLE
Distribution: Implement compounding by numerical integration

### ISSUE THREAD

**Issue by sergey-a-berezin** [stockparfait/stockparfait#119]

**Comment by sergey-a-berezin** [comment]
As the first attempt, try direct integration over the entire space by replacing `x = t / (1 - t^2)` and integrating over `t in (-1..1)`. This can be done either by iterating over the full grid, or by using a Monte Carlo integration method over the volume. There are also adaptive methods that increase the number of points in the "interesting" parts of the volume, such as `t` near 1 or -1.

For a t-distribution with parameter `a`, I believe a good choice of a variable replacement is `x = t / (1 - t^(2b))`, where `b = ceiling(a)`; it keeps the terms under integration in reasonable ranges for all values of `t`, especially in the extremes, minimizing the rounding errors.

It's also worth solving the hyperplane inequality `sum(x) < y` analytically, so only the valid points are used. Specifically, the first `N-1` variables can vary over the full range, but the last variable has a restricted range `(-inf..y-sum(x_1..x_(N-1)))`.

**Comment by sergey-a-berezin** [comment]
Correction: grid-based enumeration for large `N` can be infeasible. For instance, for `N=250` (one year worth of trading days), even 2 points per dimension results in `2^250 ~= 10^75` points, which is definitely infeasible.

Therefore, we are left with only Monte Carlo methods.

A couple of more ideas:
- Rather than computing `G(y)` directly and then differentiating, compute `G(y2) - G(y1)` directly by integrating over the space between two hyperplanes defined by `y1` and `y2`. In particular, this constrains the last `x_N` to a range `[y1 - s, ..., y2 - s]` where `s = sum(x_1, ..., x_(N-1))`. The choice of `y_i` can be e.g. the bucket boundaries, so that p.d.f. can be computed for the middle of the bucket similarly to how it's done in `Histogram`.
- When the value of the integrated function is very small, e.g. far in the tails, it may be better to compute it directly in log-space. The function itself is a product of source p.d.f.'s, so it's just a sum of logarithms, but accumulating the arithmetic average of the samples requires a special `logsum` method which adds numbers represented by their logarithms:

```
logsum(x, y) = log(exp(x) + exp(y)) = log(exp(x) * (1 + exp(y - x))) = x + log(1 + exp(y - x))
```

This formula is the most useful when `x >= y`. In particular:
- `logsum(x, y) = x` when `x - y > k` for a large enough `k`, e.g. when `exp(-k)` is beyond the required or `float64`-supported precision for mantissa;
- `logsum(x, y) = x + exp(y - x)` when `x - y` is relatively small but still greater than 1 or 2.

**Comment by sergey-a-berezin** [comment]
Another idea:
- Extend the `Histogram` to use `float64` counts, so that it can be directly reused by c.d.f.-based distribution, e.g. obtained by numerical integration.

**Comment by sergey-a-berezin** [comment]
Rather than going for a more general integration, we can simply bias random sampling while building a histogram towards the tails, and correct the bias with the appropriate weight, thus increasing the count of the appropriate bucket not by 1, but by some weight `w(x)`.

After doing the math (to be described in the docs in more detail), the algorithm is the following:
- Add `Histogram.AddWithWeight(x, w)` method to increase the `x`'s bucket count by `w`;
- Sample `t in (-0.5..0.5)` with the uniform distribution;
- Compute `x = t / (0.25 - t^2)^b`, `x' = (0.25 + (2*b - 1)*t^2) / (0.25 - t^2)^(b+1)` and `w = f(x) * x'`;
- Add `x` with the weight `w` to the histogram.

Here `f(x)` is the p.d.f. of the distribution, and `b` is a small non-negative integer which controls the amount of bias towards the tails. Normally, for fatter tails we want higher `b` (more bias towards the tails). For a t-distribution with a parameter `a > 1` as good choice seems to be `b = ceiling(9 / a)`.

**Comment by sergey-a-berezin** [comment]
Obviously, the usefulness of the above procedure should come for `N`-compounded samples with sufficiently high `N` (e.g. `N=250`). Here, a single sample `y` consists of the sum of `N` source samples `sum(x_1, ... x_N)`, the p.d.f. becomes the product `f(y) = f(x_1) * ... * f(x_N)`, and the weight `w` is the product of source weights: `w = w_1 * ... w_N` computed as above.

The appropriate `b` can be selected experimentally.

**Comment by sergey-a-berezin** [comment]
Some more thinking and experimenting later:

- The usefulness of the variable substitution is determined by how many samples we generate for the interesting values of `y`, which is the interval `[-R..R]` that covers most of the histogram buckets except, perhaps, the extreme catch-all buckets. For simplicity, however, just assume that `[-R..R]` is the buckets range.
- In that range, the best coverage is obtained by a close-to-uniform distribution. Obviously, we also have to cover the values outside of the interval, but keep their probability relatively low, or at least prevent it from approaching 100%.
- A perfect uniform distribution for `[-R..R]` is obtained by a linear variable substitution `x = R*t`, `t in [-1..1]`. To extend it to the full real axis, we adjust the substitution as follows: `x = r*t / (1 - t^2b)` where `r = R`, and `b` controls the width of the out-of-range regions, larger `b` makes the regions smaller, and hence less probable.
- The derivative is `x'(t) = r * (1 + (2*b-1)*t^2b) / (1 - t^2b)^2`, and the sample weight is `w(t) = f(x(t)) * x'(t)`.
- A compounded sample `y = sum(x_1, ..., x_N)` near zero begins to approach the normal distribution pretty quickly for `N` of a few dozen, and certainly for `N=250`. The best way to keep it close to uniform in the desired range `[-R..R]` is to ensure that MAD of `y` is around `R`. This can be achieved by using `r = sqrt(N)*R` in the substitution for `x`.
- Experimentally, the parameter `b` should equal `b = ceiling(c*sqrt(N))`; this consistently yields the out-of-range values with probability `P( |x| > R )`:
  -  c = 1: `P(|x| > R)  < 0.5`
  - c = 5: `P(|x| > R) < 0.2`
Note, that higher values of `c` may lead to loss of precision in computing the weights, and losing even 50% of the samples is not bad if it keeps the precision higher.

**Comment by sergey-a-berezin** [comment]
To implement all of this, add a `CompoundHistogramDistribution` method which computes a `HistogramDistribution` out of an `N`-compounded source distribution using the above Monte Carlo method and the source's `Prob(x)` method. Reuse `RandDistributionConfig`, perhaps renamed, for the necessary parameters. The histogram can be computed in parallel, similarly to how it's done in `RandDistribution`.

**Comment by sergey-a-berezin** [comment]
Add an error estimation to each `Histogram` bucket as the sum of absolute distances between consecutive results (deviations). The error is the mean of these deviations.

Also, update `PreciseEnough` to consider `epsilon` as a relative error when `[x-dev..x+dev]` lies entirely on one side of `0`. This, in particular, allows for correct error estimation of very small but strictly positive quantities such as histogram's weighted bucket values.

**Comment by sergey-a-berezin** [comment]
On error estimation, the most accurate way to parallelize it is to accumulate `sum(y)` (already done) and `sum(y^2)`, and the error then is the variance 

```
Var(err) = sum(y^2) / N - sum(y)^2 / N
Sigma(err) = sqrt(Var(err))
```

**Comment by sergey-a-berezin** [comment]
Can I use [Welford's online algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance) for calculating variance?
Check out one of the [references](http://cpsc.yale.edu/sites/default/files/files/tr222.pdf) on that page: it's a tech report from 1983 (later published as a paper), which describes the exactly how to do it, including parallel processing.

The gist of it:
- Accumulate
```
T_i,j = sum_{k=i..j} x_k` and
S_i,j = sum_{k=i..j} (x_k - Ti,j / (j-i+1) )^2
```
The desired `Var(X) = S_1,N / N`.

Combining two accumulators from two parallel runs:

```
T_1,m+n = T_1,m + T_m+1,m+n
S_1_m+n = S_1,m + S_m+1,m+n + m/(n*(m+n)) * (n / m * T_1,m - T_m+1,m+n)^2
```

In particular, for `n=1` (adding a single sample within each parallel job) this reduces to:
```
T_1,j = T_1,j-1 + x_j
S_1,j = S1,j-1 + 1/(j*(j-1)) * (j*x_j - T_1,j)^2
```

**Comment by sergey-a-berezin** [comment]
As a refinement, find a constant shift `d`, e.g. as the first sample `x_1`, or the mean of the first few samples, and use `x_i - d` instead of `x_i`. It preserves the variance, but further reduces the rounding errors, so much so that even the classical sum of squares can be used with reasonable precision.

**Comment by sergey-a-berezin** [comment]
It's worth creating a separate data structure with an interface which estimates errors from parallel online sampling using the above technique. This should make it easy to instrument various points of use.

**Comment by sergey-a-berezin** [comment]
When adjusting the standard error of a bucket, we should not be adding zeros for the samples outside the bucket, since in fact they may result in non-zero p.d.f. values divided by the larger number of samples. Need to figure out the correct way to adjust the errors.

**Comment by sergey-a-berezin** [comment]
Histogram error estimates still don't make sense, since for weights equal `1.0` the error comes out exactly zero according to the formula (which is wrong), while weighted histograms apparently grossly overestimate the errors.

The most accurate way would be to update p.d.f. values for every column for every bucket update. This, however, increases the complexity of `Histogram.Add` to `O(N*n)`, where `N` is the number of buckets and `n` is the number of samples.

Instead, we can run such a global error estimation update every `k*N` updates for some configurable `k >= 1`, so the amortized complexity is still `O(n)` and doesn't depend on `N`.
